src/commands/BootstrapCommand.js

```diff
@@ -129,32 +130,251 @@ export default class BootstrapCommand extends Command {
+    });
+    return { root, leaves };
+  }
+
   /**
    * Install external dependencies for all packages
    * @param {Function} callback
    */
   installExternalDependencies(callback) {
-    this.logger.info("Installing external dependencies");
-    this.progressBar.init(this.filteredPackages.length);
+    const {leaves, root} = this.getDependenciesToInstall(this.filteredPackages);
+
     const actions = [];
-    this.filteredPackages.forEach((pkg) => {
-      const allDependencies = pkg.allDependencies;
-      const externalPackages = Object.keys(allDependencies)
-        .filter((dependency) => {
-          const match = find(this.packages, (pkg) => {
-            return pkg.name === dependency;
+
+    // Start root install first, if any, since it's likely to take the longest.
+    if (Object.keys(root).length) {
+      actions.push((cb) => NpmUtilities.installInDir(
+        this.repository.rootPath,
+        root.map(({dependency}) => dependency).filter((dep) => dep),
+        (err) => {
+          if (err) return cb(err);
+
+          // Link binaries into dependent packages so npm scripts will have
+          // access to them.
+          async.series(root.map(({name, dependents}) => (cb) => {
+            const {bin} = (this.hoistedPackageJson(name) || {});
+            if (bin) {
+              async.series(dependents.map((pkg) => (cb) => {
+                const src  = this.hoistedDirectory(name);
+                const dest = pkg.nodeModulesLocation;
+                this.createBinaryLink(src, dest, name, bin, cb);
+              }), cb);
+            } else {
+              cb();
+            }
+          }), (err) => {
+            this.progressBar.tick("Install hoisted");
+            cb(err);
           });
-          return !(match && pkg.hasMatchingDependency(match));
-        })
-        .filter((dependency) => !pkg.hasDependencyInstalled(dependency))
-        .map((dependency) => dependency + "@" + allDependencies[dependency]);
-      if (externalPackages.length) {
-        actions.push((cb) => NpmUtilities.installInDir(pkg.location, externalPackages, (err) => {
+        }
+      ));
+
+      // Remove any hoisted dependencies that may have previously been
+      // installed in package directories.
+      actions.push((cb) => {
+        async.series(root.map(({name, dependents}) => (cb) => {
+          async.series(dependents.map(({nodeModulesLocation: dir}) => (cb) => {
+            FileSystemUtilities.rimraf(path.join(dir, name), cb);
+          }), cb);
+        }), (err) => {
+          this.progressBar.tick("Prune hoisted");
+          cb(err);
+        });
+      });
+    }
+
+    // Install anything that needs to go into the leaves.
+    Object.keys(leaves)
+      .map((pkgName) => ({pkg: this.packageGraph.get(pkgName).package, deps: leaves[pkgName]}))
+      .forEach(({pkg, deps}) => actions.push(
+        (cb) => NpmUtilities.installInDir(pkg.location, deps, (err) => {
           this.progressBar.tick(pkg.name);
```

Comment:
The progress bar is just initialized with the number of actions we're going to run.

```javascript
this.progressBar.init(actions.length);
```

There's a vestigial comment down there, though.  I'll clean that up.